File: src/python/sort_insertion_sort.py
import os
import sys
from typing import List


#####################
# sort script start #
#####################


# insertion sort!
#
# docs: https://en.wikipedia.org/wiki/Insertion_sort
#
# insertion sort steps through an input list, using it to grow an output list.
# on every step, it inserts the current element into its proper location on the output list.
# it continues until the input list is empty.


def do_sorting(input_list):
    return insertion_sort(input_list)


def insertion_sort(input_list: List[str]) -> List[str]:
    output_list = []

    for index, element in enumerate(input_list):
        output_list.append(element)
        target_index = index
        while (target_index != 0) and (element < output_list[index - 1]):
            # swap order
            output_list[target_index], output_list[target_index - 1] = (
                output_list[target_index - 1],
                output_list[target_index],
            )
            target_index = target_index - 1

    return output_list


###################
# sort script end #
###################

# 👇🏽 copy pasted helpers

if __name__ == "__main__":
    # read input file
    inputFilePath = os.getenv("INPUT_PATH")
    with open(inputFilePath, "r") as inputFileObject:
        inputFileData = inputFileObject.readlines()

    # clean input data
    cleanedInputData = []
    for element in inputFileData:
        cleanedInputData.append(element.strip())

    # do sorting
    sortedData = do_sorting(cleanedInputData)

    # write output file
    outputFilePath = os.getenv("OUTPUT_PATH")
    with open(outputFilePath, "w") as outputFileObject:
        for element in sortedData:
            outputFileObject.write(element + "\n")

File: src/rust/sort_insertion_sort.rs
use std::env;
use std::fs;
use std::fs::File;
use std::io::Write;

///////////////////////
// sort script start //
///////////////////////

// insertion sort!
//
// docs: https://en.wikipedia.org/wiki/Insertion_sort
//
// insertion sort steps through an input list, using it to grow an output list.
// on every step, it inserts the current element into its proper location on the output list.
// it continues until the input list is empty.

fn do_sorting(input_list: Vec<&str>) -> Vec<&str> {
  return insertion_sort(input_list);
}

fn insertion_sort(input_list: Vec<&str>) -> Vec<&str> {
  let mut output_list: Vec<&str> = vec![];

  for (index, element) in input_list.iter().enumerate() {
    if index == 0 {
      output_list.push(element);
    } else if element < &output_list[index - 1] {
      let mut target_index = index;
      while &output_list[target_index] < &output_list[target_index - 1] {
        output_list.swap(target_index, target_index - 1);
        target_index = target_index - 1;
      }
    } else {
      output_list.push(element);
    }
  }

  return output_list;
}

/////////////////////
// sort script end //
/////////////////////

// 👇🏽 copy pasted helpers

fn main() -> Result<(), Box<dyn std::error::Error + 'static>> {
  // setup
  let input_file_path = env::var("INPUT_PATH").unwrap();
  let output_file_path = env::var("OUTPUT_PATH").unwrap();

  // get input data
  let input_data_string = fs::read_to_string(input_file_path)?;
  let input_data_vec = input_data_string.lines().collect();

  // do sorting
  let sorted_data = do_sorting(input_data_vec);

  // write output data
  // join on \n, and add an additional trailing \n
  let output_data_string = sorted_data.join("\n") + "\n";
  let output_data_bytes = output_data_string.as_bytes();
  let mut output_file = File::create(output_file_path)?;
  output_file.write_all(output_data_bytes)?;

  // teardown
  Ok(())
}
